File: dairy_erp/dairy_erp/report/local_sales_report/local_sales_report.py
# Copyright (c) 2013, indictrans technologies and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _
from frappe.utils import has_common

def execute(filters=None):
	columns, data = get_columns(), get_data(filters)
	return columns ,data 

def get_columns():

	columns = [
		_("Date") + ":Data:90", 
		_("Customer") + ":Link/Customer:150",
		_("Shift") + ":Data:90",
		_("Sales Invoice") + ":Link/Sales Invoice:150",
		_("Item Code") + ":Link/Item:150",
		_("Quantity") + ":Float:150",
		_("Rate") + ":Float:150",
		_("Amount") + ":Currency:100"
	]

	return columns

def get_data(filters):

	vlcc_comp = frappe.db.get_value("User",frappe.session.user,"company")

	data = frappe.db.sql("""select DATE_FORMAT(si.posting_date, "%d-%m-%y"),
								   si.customer,
								   LEFT(si.shift,1),
								   si.name,
								   si_item.item_name,
								   si_item.qty,
								   si_item.rate,
								   si.grand_total
							from 
								`tabSales Invoice` si,
								`tabSales Invoice Item` si_item,
								`tabCustomer` cus
							where 
								si.local_sale = 1 and
								si.customer = cus.name
								and si.name = si_item.parent
								and si.docstatus = 1 and si.company = '{0}'
								{1}""".format(vlcc_comp,get_conditions(filters)),as_list=1,debug=1)
	
	if data:
		if filters.get('customer_type') == "Farmer":
			g_total = 0
			qty_total = 0
			for row in data:
				qty_total += row[5]
				g_total += row[7]
			data.append(["","","Grand Total","","",qty_total,"",g_total])		
		else:
			g_total = 0
			for row in data:
				g_total += row[7]
			data.append(["","","Grand Total","","","","",g_total])
	return data

def get_conditions(filters):
	conditions = " and 1=1"
	if filters.get('customer_type'):
		conditions += " and cus.customer_group = '{0}'".format(filters.get('customer_type'))
	if filters.get('customer'):
		conditions += " and si.customer = '{0}'".format(filters.get('customer'))
	if filters.get('from_date') and filters.get('to_date'):
		conditions += " and si.posting_date between '{0}' and '{1}' ".format(filters.get('from_date'),filters.get('to_date'))

	return conditions

def get_customer(doctype,text,searchfields,start,pagelen,filters):
	branch_office = frappe.db.get_value("User", frappe.session.user, ['branch_office','company'],as_dict=True)
	if filters.get('customer_type'):
		if has_common(frappe.get_roles(), ["Vlcc Manager", "Vlcc Operator"]) and frappe.session.user != 'Administrator':
			# return frappe.db.sql("""select name, customer_group,company
			# 					from
			# 						`tabCustomer`
			# 					where
			# 						customer_group not in ('Dairy','Vlcc') and 
			# 						name like '{text}' and company = '{comp}'""".
			# 						format(text= "%%%s%%" % text,comp = branch_office.get('company')))
			return frappe.db.sql("""select name, customer_group,company
								from
									`tabCustomer`
								where
									customer_group = '{customer_group}' and 
									name like '{text}' and company = '{comp}'""".
									format(text= "%%%s%%" % text,comp = branch_office.get('company'),customer_group = filters.get('customer_type')))
		elif frappe.session.user == 'Administrator':
			return frappe.db.sql("""select name, customer_group,company
								from
									`tabCustomer`
								where
									customer_group not in ('Dairy','Vlcc') and 
									name like '{text}' """.
									format(text= "%%%s%%" % text))
	

File: dairy_erp/dairy_erp/report/local_sales_report/local_sales_report.js
// Copyright (c) 2016, indictrans technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Local Sales Report"] = {
	"filters": [
		{
			"fieldname":"customer_type",
			"label": __("VLCC Local Customer/Farmer"),
			"fieldtype": "Select",
			"options": 'Vlcc Local Customer\nFarmer\nVlcc Local Institution',
			"default": "Vlcc Local Customer",
			"on_change": function(query_report) {
				var customer = frappe.query_report_filters_by_name.customer.get_value()
				if (customer){
					frappe.query_report_filters_by_name.customer.set_input("");
				}
				query_report.trigger_refresh();
			}
		},
		{
			"fieldname":"customer",
			"label": __("Customer"),
			"fieldtype": "Link",
			"options":"Customer",
			"get_query": function (query_report) {
				return{
					query:"dairy_erp.dairy_erp.report.local_sales_report.local_sales_report.get_customer",
					"filters": {
						"customer_type": frappe.query_report_filters_by_name.customer_type.get_value()
						}
					}
				},
			"on_change": function(query_report) {
				var customer_type = frappe.query_report_filters_by_name.customer_type.get_value()
				console.log("insdie on_change")
				if(customer_type == "Farmer"){
					console.log("insdie farmer cond")
					frappe.call({
						method: "frappe.client.get_value",
						args: {
							doctype: "Farmer",
							filters: {"full_name": frappe.query_report_filters_by_name.customer.get_value()},
							fieldname: ["name"]
						},
						callback: function(r) {
							console.log("insidr e",r.message)
							if(!r.exc && r.message && !in_list(["Administrator", "Guest"], frappe.session.user)){
								if(has_common(frappe.user_roles, ["Vlcc Operator", "Vlcc Manager"])){
									// $('body').find("[data-fieldname=vlcc]").val(r.message.company)
									frappe.query_report_filters_by_name.farmer_id.set_input(r.message.name);
								}
								query_report.trigger_refresh();
							}
						}
					})
				}
			}	
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"on_change": function(query_report) {
				var to_date = frappe.query_report_filters_by_name.to_date.get_value()
				var from_date = frappe.query_report_filters_by_name.from_date.get_value()
				if (frappe.datetime.str_to_obj(to_date) < frappe.datetime.str_to_obj(from_date)){
					frappe.throw("To date cannot be less than from date")
				}
				query_report.trigger_refresh();
			}
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"default": frappe.datetime.get_today(),
			"on_change": function(query_report) {
				var to_date = frappe.query_report_filters_by_name.to_date.get_value()
				var from_date = frappe.query_report_filters_by_name.from_date.get_value()
				if (frappe.datetime.str_to_obj(to_date) < frappe.datetime.str_to_obj(from_date)){
					frappe.throw("To date cannot be less than from date")
				}
				query_report.trigger_refresh();
			}
		},
		{
			"fieldname":"vlcc",
			"label": __("VLCC"),
			"fieldtype": "Link",
			"options":"Company",
			"read_only": 1
		},
		{
			"fieldname":"vlcc_addr",
			"label": __("VLCC Address"),
			"fieldtype": "Data",
			"hidden": 1
		},
		{
			"fieldname":"farmer_id",
			"label": __("Farmer Id"),
			"fieldtype": "Data",
			"hidden": 1
		}

	],
	onload: function(query_report) {
		frappe.call({
			method: "frappe.client.get_value",
			args: {
				doctype: "User",
				filters: {"name": frappe.session.user},
				fieldname: ["operator_type","company", "branch_office"]
			},
			callback: function(r) {
				if(!r.exc && r.message && !in_list(["Administrator", "Guest"], frappe.session.user)){
					if(has_common(frappe.user_roles, ["Vlcc Operator", "Vlcc Manager"])){
						// $('body').find("[data-fieldname=vlcc]").val(r.message.company)
						frappe.query_report_filters_by_name.vlcc.set_input(r.message.company);
					}
					query_report.trigger_refresh();
				}
			}
		})
		frappe.call({
			method: "frappe.client.get_value",
			args: {
				doctype: "Village Level Collection Centre",
				filters: {"name": frappe.boot.user.first_name},
				fieldname: ["address_display"]
			},
			callback: function(r) {
				console.log("insidr e",r.message)
				if(!r.exc && r.message && !in_list(["Administrator", "Guest"], frappe.session.user)){
					if(has_common(frappe.user_roles, ["Vlcc Operator", "Vlcc Manager"])){
						// $('body').find("[data-fieldname=vlcc]").val(r.message.company)
						frappe.query_report_filters_by_name.vlcc_addr.set_input(r.message.address_display);
					}
					query_report.trigger_refresh();
				}
			}
		})
	}
}